utils/kafka-utils/src/main/scala/pl/touk/nussknacker/engine/kafka/exception/KafkaJsonExceptionSerializationSchema.scala

```diff
@@ -14,11 +16,123 @@ import java.io.{PrintWriter, StringWriter}
 import java.lang
 import java.net.InetAddress
 import java.nio.charset.StandardCharsets
+import scala.annotation.tailrec
 import scala.io.Source
 
+object KafkaJsonExceptionSerializationSchema {
+  private[exception] val recordOverhead = DefaultRecordBatch.RECORD_BATCH_OVERHEAD
+  private val indentLength              = 2
+  private val warningKey                = "!warning"
+
+  private def serializeValueWithSizeLimit(value: KafkaExceptionInfo, maxValueBytes: Int): Array[Byte] = {
+    val valueJson       = value.asJson
+    val serializedValue = valueJson.spaces2.getBytes(StandardCharsets.UTF_8)
+    if (value.inputEvent.isDefined && serializedValue.length > maxValueBytes) {
+      removeInputEventKeys(valueJson, serializedValue.length, maxValueBytes).spaces2.getBytes(StandardCharsets.UTF_8)
+    } else {
+      serializedValue
+    }
+  }
+
+  private def removeInputEventKeys(value: Json, valueBytes: Int, maxValueBytes: Int): Json = {
+    if (valueBytes <= maxValueBytes) {
+      return value
+    }
+
+    val removedBytesPlaceholder = "$" * valueBytes.toString.length
+    // text below will increase JSON size, but we are working under the assumption that its size is much smaller
+    // than the size of inputEvent keys that will get removed
+    val messageTemplate =
+      s"inputEvent truncated, original error event had $valueBytes bytes, which was more then the max allowed length of $maxValueBytes bytes. " +
+        s"Removed variables ($removedBytesPlaceholder bytes)"
+
+    // |{
+    // |  "inputEvent" : {
+    // |    "name": ...
+    // (note: line below uses '' as quotes because Scala 2.12 can't handle escaped "")
+    val warningBytes = 2 * indentLength + Utils.utf8Length(s"'$warningKey' : ${messageTemplate.asJson.spaces2},\n")
+    val bytesToCut   = valueBytes + warningBytes - maxValueBytes
+
+    val variablesWithLength               = countVariableLengths(value.hcursor.downField("inputEvent"))
+    val (variablesToRemove, removedBytes) = calculateKeysToRemove(variablesWithLength, bytesToCut)
+
+    if (removedBytes <= warningBytes) {
+      // this may happen only when doing tests with really low size limits
+      return value
+    }
+
+    val finalMessage = messageTemplate.replace(removedBytesPlaceholder, removedBytes.toString) +
```

Comment:
We could have a message template as a format for the `String.format` purpose and fill it with some placeholders for count `bytesToCut` and for final message preparation. Thanks to that we would avoid tricky dollars replacement and the template would show the full message including variable names.

Reply:
I decided to do it this way (non-standard placeholder + concatenating strings) because then i don't have to do more math on the amount of space reserved for the warning message. If I used `String.format` placeholders then I would have to account for them in `warningBytes`, and I didn't want to make this code even more complicated.